js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,385 @@
+       .fetch();
+ 
+  } 
+};
+
+PipelineFixed.prototype.fetchFirstSegment = function(baseInterest)
+{
+  var interest = new Interest(baseInterest);
+  interest.setMustBeFresh(true);
+
+  this.fetchSegment(interest);
+};
+
+PipelineFixed.prototype.fetchNextSegments = function
+  (originalInterest, dataName)
+{
+  if (this.firstSegmentIsReceived === false) {
+    console.log("First segment is not received yet");
+    return;
+  }
+
+  var interest = new Interest(originalInterest);
+  // Changing a field clears the nonce so that a new nonce will be generated
+  interest.setMustBeFresh(false);
+
+  while (this.nextSegmentToRequest <= this.finalBlockId && this.segmentsOnFly <= this.windowSize) {
+    // Start with the original Interest to preserve any special selectors.
+    interest.setName(dataName.getPrefix(-1).appendSegment(this.nextSegmentToRequest));
+    interest.refreshNonce();
+
+    this.fetchSegment(interest);
+
+    this.nextSegmentToRequest += 1;
+    this.increaseNumberOfSegmentsOnFly();
+  }
+};
+
+PipelineFixed.prototype.cancelPendingInterestsAboveFinalBlockId = function ()
+{
+  var len = this.dataFetchersContainer.length;
+  var i = this.finalBlockId;
+
+  for (var i = this.finalBlockId; i < len; i++) {
+    if (this.dataFetchersContainer[i] !== null) {
+      this.dataFetchersContainer[i].cancelPendingInterest();
+    }
+  }
+};
+
+PipelineFixed.prototype.onData = function(originalInterest, data)
+{
+  if (this.validatorKeyChain != null) {
+    try {
+      var thisPipeline = this;
+      this.validatorKeyChain.verifyData
+        (data,
+         function(localData) {
+           thisPipeline.onVerified(localData, originalInterest);
+         },
+         this.onValidationFailed.bind(this));
+    } catch (ex) {
+      console.log("Error in KeyChain.verifyData: " + ex);
+    }
+  }
+  else {
+    if (!this.verifySegment(data)) {
+      this.reportError(PipelineFixed.ErrorCode.SEGMENT_VERFIFICATION_FAILED,
+                       "Segment verification failed");
+    }
+
+    this.onVerified(data, originalInterest);
+  }
+};
+
+PipelineFixed.prototype.onVerified = function(data, originalInterest)
+{
+  var currentSegment = 0;
+
+  if (!PipelineFixed.endsWithSegmentNumber(data.getName())) {
+    // We don't expect a name without a segment number.  Treat it as a bad packet.
+    this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+                     "Got an unexpected packet without a segment number: " +
+                      data.getName().toUri());
+  }
+  else {
+    try {
+      currentSegment = data.getName().get(-1).toSegment();
+    }
+    catch (ex) {
+      this.reportError(PipelineFixed.ErrorCode.DATA_HAS_NO_SEGMENT,
+                       "Error decoding the name segment number " +
+                       data.getName().get(-1).toEscapedString() + ": " + ex);
+      return;
+    }
+  }
+
+  // if segment number of the very first Data packet does not start
```

Comment:
There's no reason to special case segment 0. Whatever segment was retrieved by the initial Interest, save it in `contentParts`, and skip that segment number when sending subsequent Interests.

Reply:
I do save the received segment. There are 2 scenarios:
1- the very first received segment is zero => then save the segment and continue with asking for segment 1 and so on ...
2- the very first received segment is not zero  => then save the segment and continue with asking for segment 0, 1, so on ... (we do not re-issue an interest for existing segments (look `fetchNextSegments` method).

Reply:
These are the same scenario:

1. Save whatever segment you receive from the initial Interest.
2. Express up to M Interests, skipping the segment you already have.

Thus:

* `firstSegmentIsReceived` is redundant.
* step3-5 needs to be rewritten.